core/src/main/java/com/radixdlt/consensus/vertexstore/VertexStoreJavaImpl.java

```diff
@@ -302,45 +340,88 @@ public InsertVertexChainResult insertVertexChain(VertexChain vertexChain) {
+    }
+    // ...all good (size-wise), let's continue the insertion process.
 
-  private Option<BFTInsertUpdate> insertVertexInternal(VertexWithHash vertexWithHash) {
-    LinkedList<ExecutedVertex> previous =
-        getPathFromRoot(vertexWithHash.vertex().getParentVertexId());
+    final var previous = getPathFromRoot(vertexWithHash.vertex().getParentVertexId());
     final var executedVertexOption = Option.from(ledger.prepare(previous, vertexWithHash));
-    return executedVertexOption.map(
-        executedVertex -> {
-          vertices.put(executedVertex.getVertexHash(), executedVertex.getVertexWithHash());
-          executedVertices.put(executedVertex.getVertexHash(), executedVertex);
-          vertexChildren.put(executedVertex.getVertexHash(), new HashSet<>());
-          Set<HashCode> siblings = vertexChildren.get(executedVertex.getParentId());
-          siblings.add(executedVertex.getVertexHash());
-
-          VertexStoreState vertexStoreState = getState();
-          return BFTInsertUpdate.insertedVertex(executedVertex, siblings.size(), vertexStoreState);
-        });
+    return executedVertexOption
+        .<Result<BFTInsertUpdate, VertexInsertError>>map(
+            executedVertex -> {
+              // The vertex was executed successfully, so we're inserting it
+              vertices.put(executedVertex.getVertexHash(), executedVertex.getVertexWithHash());
+              executedVertices.put(executedVertex.getVertexHash(), executedVertex);
+              vertexChildren.put(executedVertex.getVertexHash(), new HashSet<>());
+              Set<HashCode> siblings = vertexChildren.get(executedVertex.getParentId());
+              siblings.add(executedVertex.getVertexHash());
+
+              // We've already calculated the post-insert state (and verified
+              // its size against the limit), so we can just use it here.
+              trackCurrentStateSize(postInsertSerializedState);
+
+              // Update the metrics
+              metrics.bft().vertexStore().size().set(vertices.size());
+              if (siblings.size() > 1) {
+                metrics.bft().vertexStore().forks().inc();
+              }
+              if (!vertexWithHash.vertex().hasDirectParent()) {
+                metrics.bft().vertexStore().indirectParents().inc();
+              }
+
+              return Result.success(new BFTInsertUpdate(executedVertex, postInsertSerializedState));
+            })
+        .orElse(Result.error(VertexInsertError.PREPARE_FAILED));
   }
 
-  private void removeVertexAndPruneInternal(HashCode vertexId, HashCode skip) {
-    vertices.remove(vertexId);
+  private void removeVertexAndPruneInternal(HashCode vertexId, Optional<HashCode> skip) {
+    Optional.ofNullable(vertices.remove(vertexId))
+        .flatMap(
+            removedVertex ->
+                Optional.ofNullable(vertexChildren.get(removedVertex.vertex().getParentVertexId())))
+        .ifPresent(siblings -> siblings.remove(vertexId));
```

Comment:
some lines could be a lot less awkward if `this.vertexChildren` was really a `Multimap<HashCode, HashCode>` (e.g. `= HashMultimap.create()`)

Reply:
Splendid idea! Done.